lib/collection/src/collection/shard_transfer.rs

```diff
@@ -257,33 +256,37 @@ impl Collection {
             .stop_task(&transfer_key)
             .await;
 
-        let replica_set =
-            if let Some(replica_set) = shard_holder_guard.get_shard(&transfer_key.shard_id) {
-                replica_set
-            } else {
-                return Err(CollectionError::bad_request(format!(
-                    "Shard {} doesn't exist",
-                    transfer_key.shard_id
-                )));
-            };
+        let mut shard_holder_guard = None;
+
+        let shard_holder = match shard_holder {
+            Some(shard_holder) => shard_holder,
+            None => shard_holder_guard.insert(self.shards_holder.read().await),
+        };
 
-        let transfer = shard_holder_guard.get_transfer(&transfer_key);
+        let Some(replica_set) = shard_holder.get_shard(&transfer_key.shard_id) else {
+            return Err(CollectionError::bad_request(format!(
+                "Shard {} doesn't exist",
+                transfer_key.shard_id
+            )));
+        };
+
+        let Some(transfer) = shard_holder.get_transfer(&transfer_key) else {
+            return Ok(());
+        };
 
-        if transfer.map(|x| x.sync).unwrap_or(false) {
-            replica_set.set_replica_state(&transfer_key.to, ReplicaState::Dead)?;
-        } else {
-            replica_set.remove_peer(transfer_key.to).await?;
+        if replica_set.peer_state(&transfer.to).is_some() {
+            if transfer.sync {
+                replica_set.set_replica_state(&transfer.to, ReplicaState::Dead)?;
+            } else {
+                replica_set.remove_peer(transfer.to).await?;
+            }
         }
 
-        if self.this_peer_id == transfer_key.from {
-            transfer::driver::revert_proxy_shard_to_local(
-                shard_holder_guard,
-                transfer_key.shard_id,
-            )
-            .await?;
+        if transfer.from == self.this_peer_id {
+            transfer::driver::revert_proxy_shard_to_local(shard_holder, transfer.shard_id).await?;
         }
 
-        let _finish_was_registered = shard_holder_guard.register_finish_transfer(&transfer_key)?;
+        shard_holder.register_finish_transfer(&transfer_key)?;
```

Comment:
Seems safe to only do this when `transfer.is_some()`, because this just removes the transfer from the local state, so if `transfer = None` already, it's just a no-op.